### PR TITLE
Allow arbitrary env variables in bindings, use process.env

### PR DESCRIPTION
**Pull request by apedroferreira**
- Allow for typing in any environment variable name in environment variable bindings, even if it doesn't exist
- Correctly retrieve environment variable value from `process.env`
- Refresh `process.env` when env file changes

https://github.com/mui/mui-toolpad/assets/10789765/ea9eff13-a372-4210-9d3f-fea205b0e3bc

Fixes 1 and 2 in https://github.com/mui/mui-toolpad/issues/2099.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -166,6 +166,7 @@ jobs:
           name: Tests real browser
           environment:
             HTTPBIN_BASEURL: 'http://localhost/'
+            TEST_ENV_VAR: 'foo'
```
Comment (Janpot) [inline .circleci/config.yml:169]: 👍 but I'd keep it local like in https://github.com/mui/mui-toolpad/pull/2103/files#diff-f16a8b4fc36e8dc38f1edc79e02791f23fed2ae1be7d7f7b1b015c1a05aec6ebR23-R25

Comment (apedroferreira) [reply .circleci/config.yml:169]: Ah so there is a way, I was looking into this right now, thanks!

File: test/integration/rest-basic/index.spec.ts [hunk 2]
```diff
@@ -41,14 +44,6 @@ test('rest basics', async ({ page, context, localApp }) => {
 
   await expect(page.getByText('query4 authorization: foo')).toBeVisible();
 
-  const envFilePath = path.resolve(localApp.dir, './.env');
```
Comment (Janpot) [inline test/integration/rest-basic/index.spec.ts:44]: Why was this removed? This is testing whether the editor detects changes to the env file correctly

**Review by Janpot — CHANGES_REQUESTED** [review]
We should keep the test with `.env` file and add another one with a different variable for the behavior of environemnt variables not in the `.env` file.

It's important because we need to test that changes to the file correctly propagate

File: test/integration/rest-basic/index.spec.ts (hunk 2 above)
Comment (apedroferreira) [reply test/integration/rest-basic/index.spec.ts:44]: Dotenv wasn't working with tests locally, I'll try to make it work though so we can test like this.

Comment (Janpot) [reply test/integration/rest-basic/index.spec.ts:44]: Both methods need to work. The key is to add a second env var so that the old test remains as it is. Like it's done for https://github.com/mui/mui-toolpad/pull/2103/files#diff-f16a8b4fc36e8dc38f1edc79e02791f23fed2ae1be7d7f7b1b015c1a05aec6ebR23-R25

File: packages/toolpad-app/src/server/EnvManager.ts [hunk 3]
```diff
@@ -46,6 +46,8 @@ export default class EnvManager {
     envFileWatcher.on('all', async () => {
       // Invalidate cache
       this.values = undefined;
+
+      dotenv.config({ override: true });
```
Comment (apedroferreira) [inline packages/toolpad-app/src/server/EnvManager.ts:50]: This here refreshes `process.env` to match the env file when it changes.

**Review by Janpot — APPROVED** [review]
Not a huge fan of updating `process.env` at runtime. Most code in the repo and in libraries expect `process.env` to be static throughout the app lifecycle (e.g. look at our src/config.ts, those values don't update when process.env changes). It'd probably be more maintainable to construct the env explicitly when interpolating `$$env` bindings.
Happy to merge though if you disagree.

**Comment by apedroferreira** [comment]
> Not a huge fan of updating process.env at runtime. Most code in the repo and in libraries expect process.env to be static throughout the app lifecycle (e.g. look at our src/config.ts, those values don't update when process.env changes). It'd probably be more maintainable to construct the env explicitly when interpolating $$env bindings.

So you'd prefer we read the `.env` file manually in development?
Or do you mean we should read the whole yaml files and list every `$$env` binding that is there?
I thought this solution would be more solid but if there's other things depending on `process.env` being static I don't think the solution is worse.

**Comment by apedroferreira** [comment]
I might also merge now but then will add that change asap, just so that the feature works.
